DESIGN.md

```diff
@@ -0,0 +1,79 @@
+
+To be hypervisor neutral, the high level abstraction has been heavily refactored. The new `vm-memory` crate could be divided into four logic parts as:
+
+### Abstraction of Generic Address Space 
+Build generic abstractions to describe and access an address space as below:
+- AddressValue: Stores the raw value of an address. Typically u32, u64 or usize is used to store the raw value. But pointers, such as \*u8, can't be used because it doesn't implement the Add and Sub traits.
+- Address: Encapsulates an AddressValue object and defines methods to access it.
+- Bytes: Common trait for volatile access to memory. The `Bytes` trait can be parameterized with newtypes that represent addresses, in order to enforce that addresses are used with the right "kind" of volatile memory.
+- VolatileMemory: Basic implementation of volatile access to memory, implements `Bytes<usize>`.
+
+To make the abstraction as generic as possible, all of above core traits only define methods to access the address space, and they never define methods to manage (create, delete, insert, remove etc) address spaces. By this way, the address space consumers (virtio device drivers, vhost-user drivers and boot loaders etc) may be decoupled from the address space provider (typically a hypervisor).
+
+### Specialization for Virtual Machine Physical Address Space
+The generic address space crates are specialized to access VM's physical memory with following traits:
+- GuestAddress: represents a guest physical address (GPA). On ARM64, a 32-bit hypervisor may be used to support a 64-bit VM. For simplicity, u64 is used to store the the raw value no matter it is a 32-bit or 64-bit virtual machine.
+- GuestMemoryRegion: used to represent a continuous region of VM's physical memory.
+- GuestMemory: used to represent a collection of GuestMemoryRegion objects. The main responsibilities of the GuestMemory trait are:
+	- hide the detail of accessing VM's physical address (for example complex hierarchical structures).
+	- map a request address to a GuestMemoryRegion object and relay the request to it.
+	- handle cases where an access request spanning two or more GuestMemoryRegion objects.
+
+The VM memory consumers, such as virtio device drivers, vhost drivers and boot loaders etc, should only rely on traits and structs defined here to access VM's physical memory.
+
+### A Sample and Default Backend Implementation Based on mmap()
+Provide a default and sample implementation of the GuestMemory trait by mmapping VM's physical memory into current process. Three data structures are defined here:
+- MmapRegion: mmap a continous range of VM's physical memory into current and provide methods to access the mmapped memory.
+- GuestRegionMmap: a wrapper structure to map VM's physical address into (mmap\_region, offset) tuple.
+- GuestMemoryMmap: manage a collection of GuestRegionMmap objects for a VM.
```

Comment:
I think here we want to highlight the fact that `GuestMemoryMmap` is the `GuestMemory` implementation for an mmap'ed volatile memory backend.
That's our top level VM memory interface for the mmap backend, and we should highlight that.

I'm thinking about adding a sub-section to this section with really simple examples on how  to use `GuestMemoryMmap`. I'll do so as a follow-up PR.